saml-core/src/test/java/org/keycloak/saml/processing/core/saml/v2/writers/XUATokenWriterTest.java

```diff
@@ -0,0 +1,62 @@
+
+import java.io.ByteArrayOutputStream;
+
+import org.junit.Assert;
+import org.junit.Test;
+import org.keycloak.dom.saml.v2.assertion.AttributeStatementType;
+import org.keycloak.dom.saml.v2.assertion.AttributeType;
+import org.keycloak.saml.common.constants.GeneralConstants;
+import org.keycloak.saml.common.constants.JBossSAMLURIConstants;
+import org.keycloak.saml.common.exceptions.ConfigurationException;
+import org.keycloak.saml.common.exceptions.ProcessingException;
+import org.keycloak.saml.common.util.DocumentUtil;
+import org.keycloak.saml.common.util.StaxUtil;
+import org.w3c.dom.Attr;
+import org.w3c.dom.Document;
+import org.w3c.dom.Element;
+
+public class XUATokenWriterTest {
+
+    @Test
+    public void testXUAToken() throws ConfigurationException, ProcessingException {
+        Document document = DocumentUtil.createDocument();
+
+        AttributeType roleAttr = new AttributeType("urn:oasis:names:tc:xacml:2.0:subject:role");
+        Element attributeValue =
+                document.createElementNS(JBossSAMLURIConstants.ASSERTION_NSURI.get(), "AttributeValue");
+        Attr attrRoleType = document.createAttributeNS(JBossSAMLURIConstants.XSI_NSURI.get(), "type");
+        attrRoleType.setValue("Role");
+        attrRoleType.setPrefix("xsi");
+        attributeValue.setAttributeNodeNS(attrRoleType);
+        Element role = document.createElementNS("urn:hl7-org:v3", "Role");
+        role.setAttributeNS("urn:hl7-org:v3", "code", "46255001");
+        role.setAttributeNS("urn:hl7-org:v3", "codeSystem", "2.16.840.1.113883.6.96");
+        role.setAttributeNS("urn:hl7-org:v3", "codeSystemName", "SNOMED_CT");
+        role.setAttributeNS("urn:hl7-org:v3", "displayName", "Pharmacist");
+        Attr attrCEType = document.createAttributeNS(JBossSAMLURIConstants.XSI_NSURI.get(), "type");
+        attrCEType.setValue("CE");
+        attrCEType.setPrefix("xsi");
+        role.setAttributeNodeNS(attrCEType);
+        attributeValue.appendChild(role);
+        roleAttr.addAttributeValue(attributeValue);
+
+        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
+        SAMLAssertionWriter samlAssertionWriter =
+                new SAMLAssertionWriter(StaxUtil.getXMLStreamWriter(byteArrayOutputStream));
+
+        AttributeStatementType attributeStatementType = new AttributeStatementType();
+        attributeStatementType.addAttribute(new AttributeStatementType.ASTChoiceType(roleAttr));
+
+        samlAssertionWriter.write(attributeStatementType);
+
+        String serializedAssertion = new String(byteArrayOutputStream.toByteArray(), GeneralConstants.SAML_CHARSET);
+        Assert.assertEquals("<saml:AttributeStatement><saml:Attribute Name=\"urn:oasis:names:tc:xacml:2"
+                        + ".0:subject:role\"><AttributeValue xmlns=\"urn:oasis:names:tc:SAML:2.0:assertion\" "
+                        + "xsi:type=\"Role\"><Role xmlns=\"urn:hl7-org:v3\" code=\"46255001\" "
```

Comment:
extra attribute in AttributeValue does not conform with SAML schema
```suggestion
                        + "><Role xmlns=\"urn:hl7-org:v3\" code=\"46255001\" "
```

Reply:
I think it does conform - because for String values _xsi:type="xs:string"_ is written.